crates/autopilot/src/domain/settlement/mod.rs

```diff
@@ -40,15 +41,55 @@ impl Settlement {
 
         if persistence
             .auction_has_settlement(solution.auction_id())
-            .await?
+            .await
+            .map_err(Error::from)?
         {
             // This settlement has already been processed by another environment.
             //
             // TODO: remove once https://github.com/cowprotocol/services/issues/2848 is resolved and ~270 days are passed since bumping.
             return Err(Error::WrongEnvironment);
         }
 
-        let auction = persistence.get_auction(solution.auction_id()).await?;
+        let auction = persistence
+            .get_auction(solution.auction_id())
+            .await
+            .map_err(Error::from)?;
+
+        let competition_winner = persistence
+            .get_competition_winner(solution.auction_id())
+            .await
+            .map_err(Error::from)?;
+
+        if transaction.solver != competition_winner.solver() {
+            return Err(Error::WinnerMismatch {
+                expected: competition_winner.solver(),
+                got: transaction.solver,
+            });
+        }
+
+        // only debugging for now
+        match solution.score(&auction) {
+            Ok(score) => {
+                tracing::debug!(
+                    "Settlement for auction {} has score {}",
+                    solution.auction_id(),
+                    score
+                );
+
+                if score < competition_winner.score() {
+                    tracing::warn!(
+                        "Settlement for auction {} has lower score {} than the competition winner \
+                         {}",
+                        solution.auction_id(),
+                        score,
+                        competition_winner.score()
+                    );
+                }
+            }
+            Err(err) => {
+                tracing::warn!(?err, "failed to calculate score for settlement");
+            }
+        }
```

Comment:
Why not turn this into a proper error. You are already logging the result of `Settlement::new` somewhere else for debugging purposes. Logging it here again just adds more places we are doing things temporarily for testing. The way you had it before was better imo.

Reply:
I'm logging settlement observation only. Score is not logged and it's impossible to log since it's part of the settlement constructor now. I was hesitant of returning error before making sure the score is properly calculated through debugging.

Reply:
you were logging it before though (you can still log if the expected score doesn't match the computed one)

Reply:
yes, I had a public function `score` which is removed as not used anymore.